aten/src/ATen/Parallel_native.h

```diff
@@ -0,0 +1,174 @@
+#pragma once
+#include <ATen/ATen.h>
+#include <ATen/core/ivalue.h>
+#include <ATen/Parallel.h>
+
+#include <c10/core/thread_pool.h>
+
+#include <algorithm>
+#include <cstddef>
+#include <exception>
+
+#define INTRA_OP_PARALLEL
+
+namespace at {
+namespace internal {
+// internal function to get access to intra-op thread pool from
+// template parallel primitives (parallel_for, parallel_reduce)
+CAFFE2_API TaskThreadPoolBase& _get_intraop_pool();
+
+// internal utility function to mark master thread as in parallel
+// region when executing parallel primitives
+CAFFE2_API void _set_in_parallel_region(bool);
+
+// Simulate OMP's omp_get_thread_num() by force-setting thread local
+// task id as thread number when executing parallel primitives
+CAFFE2_API void _set_thread_num(size_t thread_num);
+CAFFE2_API void _unset_thread_num();
+}
+
+template <class F>
+void parallel_for(
+    const int64_t begin,
+    const int64_t end,
+    const int64_t grain_size,
+    const F& f) {
+  if (begin >= end) {
+    return;
+  }
+
+  if (grain_size < 0) {
+    throw std::runtime_error("Invalid begin, end or grain_size in parallel_for");
+  }
+
+  if (((end - begin) >= grain_size) && !in_parallel_region()) {
+    // choose number of tasks based on grain size and number of threads;
+    size_t chunk_size = divup((end - begin), get_num_threads());
+    // make sure each task is at least grain_size size
+    chunk_size = std::max((size_t)grain_size, chunk_size);
+    size_t num_tasks = divup((end - begin), chunk_size);
+
+    std::atomic_flag err_flag = ATOMIC_FLAG_INIT;
+    std::exception_ptr eptr;
+
+    auto task = [&](int64_t task_id, int64_t local_start, int64_t local_end) {
+      internal::_set_thread_num(task_id);
+      try {
+        f(local_start, local_end);
+      } catch (...) {
+        if (!err_flag.test_and_set()) {
+          eptr = std::current_exception();
+        }
+      }
+      internal::_unset_thread_num();
+    };
+
+    std::vector<ivalue::Future> futures(num_tasks);
+    for (size_t task_id = 1; task_id < num_tasks; ++task_id) {
+      int64_t local_start = begin + task_id * chunk_size;
+      if (local_start < end) {
+        int64_t local_end = std::min(end, (int64_t)(chunk_size + local_start));
+        internal::_get_intraop_pool().run(
+            // copy task_id, local_start, local_end
+            [&futures, &task, task_id, local_start, local_end]() {
+          task(task_id, local_start, local_end);
+          futures[task_id].markCompleted(IValue());
+        });
+      } else {
+        futures[task_id].markCompleted(IValue());
+      }
+    }
+
+    int64_t first_task_end = std::min(end, (int64_t)(chunk_size + begin));
+    internal::_set_in_parallel_region(true);
```

Comment:
don't you need to set it also on child threads? (i thought that's omp's semantics)

in that case you can also just use get_thread_num

Reply:
we do return true for child threads because we call `internal::_get_intraop_pool().inThreadPool` - but i'll also set `internal::_set_in_parallel_region(true)` for them, then we probably don't need `internal::_get_intraop_pool().inThreadPool`